### PR TITLE
replace nose by pytest

### LINKED ISSUE #364: Switch to pytest/nose2/unittest
Hello,
just a heads-up that nose will no longer work with python 3.9/3.10+ :)

### PR DESCRIPTION
**Pull request by pgajdos**
Let us try to resolve #364 .

### CODE REVIEW

**Comment by chfw** [comment]
please run this:

```
pip install -r tests/requirements.txt
make format
```

if you are doing inside windows, please see format.sh for more info

Then your PR will go pass 'lint' stage.

File: .gitignore [hunk 1]
```diff
@@ -48,7 +48,6 @@ htmlcov/
 .coverage
 .coverage.*
 .cache
-nosetests.xml
```
Comment (chfw) [inline .gitignore]: don't worry about this one. 

it is automatically governed by pypi-mobans.

**Comment by chfw** [comment]
both windows and macos unit tests failed in the same way. looks like the plugin setup call were not called.

**Comment by pgajdos** [comment]
I'll look at it more. In case I have troubles, I will ask.

File: tests/integration_tests/test_command_line_options.py [hunk 2]
```diff
@@ -3,8 +3,8 @@
 from shutil import copyfile
 
 from mock import MagicMock, patch
-from nose import SkipTest
```
Comment (mcepl) [inline tests/integration_tests/test_command_line_options.py:6]: Just it is not necessary to use pytest here. `unittest` has `SkipTest` as well.

**Comment by pgajdos** [comment]
@chfw, not sure now about the failure.
https://travis-ci.org/github/moremoban/moban/jobs/729019453

**Comment by chfw** [comment]
That failed to auto generate contributor list. I need to have a detailed look. Clearly it is not related to your change.

Can you try ‘make update’?

**Comment by chfw** [comment]
it is related to project meta data update. please run `make update` or `moban -m monbanfile` and make a commit. please note:

1. some of your changes will be overrided, such as your change in gitignore
2. please use latest `gease`

the purpose of the moban task is to keep project meta info up-to-date.

**Comment by chfw** [comment]
and travis job failure seems intermittent because I could not reproduce. after rerun, it succeeded in telling why it failed.

**Comment by pgajdos** [comment]
I get following failure:

```
$ moban -m mobanfile
Info: git clone git://github.com/moremoban/pypi-mobans.git
Info: checking out submodule
Info: git clone git://github.com/moremoban/moban-anyconfig.git
Info: git clone git://github.com/moremoban/moban-handlebars.git
Info: git clone git://github.com/moremoban/moban-velocity.git
Info: git clone git://github.com/moremoban/moban-slim.git
Info: git clone git://github.com/moremoban/httpfs.git
Info: git clone git://github.com/moremoban/gitfs2.git
Info: git clone git://github.com/moremoban/pypifs.git
Templating moban_readme.jj2 to README.rst
Traceback (most recent call last):
  File "/usr/local/bin/moban", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/moban/main.py", line 50, in main
    count = handle_moban_file(moban_file, options)
  File "/usr/local/lib/python3.8/site-packages/moban/main.py", line 208, in handle_moban_file
    mobanfile.handle_moban_file_v1(moban_file_configurations, options)
  File "/usr/local/lib/python3.8/site-packages/moban/core/mobanfile/__init__.py", line 83, in handle_moban_file_v1
    number_of_templated_files = handle_targets(merged_options, targets)
  File "/usr/local/lib/python3.8/site-packages/moban/core/mobanfile/__init__.py", line 126, in handle_targets
    engine.render_to_files(jobs_for_each_engine[template_type])
  File "/usr/local/lib/python3.8/site-packages/moban/core/moban_factory.py", line 217, in render_to_files
    self._render_with_finding_data_first(sta.data_file_index)
  File "/usr/local/lib/python3.8/site-packages/moban/core/moban_factory.py", line 274, in _render_with_finding_data_first
    flag = self.apply_template(
  File "/usr/local/lib/python3.8/site-packages/moban/core/moban_factory.py", line 161, in apply_template
    rendered_content = self.engine.apply_template(
  File "/usr/local/lib/python3.8/site-packages/moban/plugins/jinja2/engine.py", line 164, in apply_template
    rendered_content = template.render(**data)
  File "/usr/local/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/usr/local/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/local/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "moban_setup.py.jj2", line 1, in top-level template code
  File "/usr/local/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/local/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "setup.py.jj2", line 413, in template
jinja2.exceptions.TemplateAssertionError: no test named 'exists'
$
```

`gease` version:

$ pip show gease
Name: gease
Version: 0.0.8
Summary: simply makes a git release using github api v3
Home-page: https://github.com/moremoban/gease
Author: C. W.
Author-email: wangc_2011@hotmail.com
License: MIT
Location: /usr/local/lib/python3.8/site-packages
Requires: crayons, requests
Required-by: moban-jinja2-github
$
```

**Comment by chfw** [comment]
for missing 'exsits', please do: `pip install moban-ansible`.

**Comment by pgajdos** [comment]
Few failures left.

File: tests/deprecated/test_repo.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
 
 
 @patch("appdirs.user_cache_dir", return_value="root")
-@patch("moban.utils.mkdir_p")
+@patch("moban.core.utils.mkdir_p")
```
Comment (chfw) [inline tests/deprecated/test_repo.py]: This bug was hidden for a while now.

Comment (chfw) [reply tests/deprecated/test_repo.py]: It is strange that nose tests did not fail.

Comment (pgajdos) [reply tests/deprecated/test_repo.py]: Thanks! There are more such places, I do not know about pytest/nose enough to tell why it has not manifested earlier.

**Comment by pgajdos** [comment]
Not sure about this one.

**Comment by pgajdos** [comment]
`fs.errors.InvalidCharsInPath: path 'tests\fixtures\issue_126\multi-key-config.yaml' contains invalid characters`

**Comment by chfw** [comment]
because moban use pyfilesystem2, 'tests\fixtures\issue_126\multi-key-config.yaml' is invalid because of '\\'. it should be '/'

**Comment by pgajdos** [comment]
Not everywhere it seems? So all os.path.join should be replaced by fs.path.join?

**Comment by pgajdos** [comment]
`moban.exceptions.PassOn: no test named 'exists'`

but `moban-ansible` is there

**Comment by pgajdos** [comment]
`fs.errors.InvalidCharsInPath: path '{{'a'|url_join('b')}} {{'b' is exists}}{{ description }}' contains invalid characters`

```
020-09-23T14:22:18.5046717Z         test_args = [
2020-09-23T14:22:18.5047234Z             "moban",
2020-09-23T14:22:18.5047756Z             "-e",
2020-09-23T14:22:18.5048403Z             "jinja2=filter:moban.externals.file_system.url_join",
2020-09-23T14:22:18.5049213Z             "jinja2=test:moban.externals.file_system.exists",
2020-09-23T14:22:18.5050086Z             "jinja2=global:description=moban.constants.PROGRAM_DESCRIPTION",
2020-09-23T14:22:18.5050785Z             "-t",
2020-09-23T14:22:18.5051456Z             "{{'a'|url_join('b')}} {{'b' is exists}}{{ description }}",
2020-09-23T14:22:18.5052083Z         ]
```
It seems that
`"{{'a'|url_join('b')}} {{'b' is exists}}{{ description }}"`
should expand to something but it does not.

**Comment by chfw** [comment]
"{{'a'|url_join('b')}} {{'b' is exists}}{{ description }}" should have been treated as a template buffer , rather than a file name , leading to a template file.

**Comment by chfw** [comment]
the 'exists' in test refers to : "jinja2=test:moban.externals.file_system.exists",

**Comment by chfw** [comment]
> "{{'a'|url_join('b')}} {{'b' is exists}}{{ description }}" should have been treated as a template buffer , rather than a file name , leading to a template file.

only when 'moban' command sees a local '.moban.yml' file under current working directory, it would mistakenly treat template buffer and template file name.

**Comment by chfw** [comment]
that's why I have to name `moban`'s .moban.yml as `mobanfile`

**Comment by pgajdos** [comment]
However, it is not apparent to me what should I do to fix these two failures.

**Comment by chfw** [comment]
yep, make sense. let me have a look at two failures on windows build. it has been a long way. thanks a lot. @pgajdos

**Review by chfw — APPROVED** [review]
Approved

**Comment by pgajdos** [comment]
@chfw, thanks, in case I can help somehow, let me know.

**Pull request merged**